src/ray/core_worker/lease_policy.cc

```diff
@@ -0,0 +1,61 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/core_worker/lease_policy.h"
+
+namespace ray {
+
+rpc::Address LocalityAwareLeasePolicy::GetBestNodeForTask(const TaskSpecification &spec) {
+  if (auto node_id = GetBestNodeIdForTask(spec)) {
+    if (auto addr = node_addr_factory_(node_id.value())) {
+      return addr.value();
+    }
+  }
+  return fallback_rpc_address_;
+}
+
+/// Criteria for "best" node: The node with the most object bytes (from object_ids) local.
+absl::optional<NodeID> LocalityAwareLeasePolicy::GetBestNodeIdForTask(
+    const TaskSpecification &spec) {
+  const auto object_ids = spec.GetDependencyIds();
+  // Number of object bytes (from object_ids) that a given node has local.
+  absl::flat_hash_map<NodeID, uint64_t> bytes_local_table;
+  uint64_t max_bytes = 0;
+  absl::optional<NodeID> max_bytes_node;
+  // Finds the node with the maximum number of object bytes local.
+  for (const ObjectID &object_id : object_ids) {
+    if (auto locality_data = locality_data_provider_->GetLocalityData(object_id)) {
+      for (const NodeID &node_id : locality_data->nodes_containing_object) {
```

Comment:
Btw, what's going to happen if the locality data is nullopt?

Reply:
This inner for loop will only be reached if the prior `if (...)` condition is true, and that [will only happen if `locality_data` contains a value](https://github.com/abseil/abseil-cpp/blob/e7ca23acac146b10edc4f752edd0bd28b0f14ea3/absl/types/optional.h#L442-L452). I've seen this `optional` pattern used and promoted elsewhere., but do you think this is too implicit? We could always opt for the more verbose
```c++
auto locality_data = locality_data_provider_->GetLocalityData(object_id);
if (locality_data.has_value()) {
    // ....
}
```